Clean up CATracking initialization (move command string parsing into workflow for now)

davidrohr opened the pull request:
This PR cleans up the initialization interface of the CATracking.
For compatibility reason, I have moved the command string parser into the workflowSpec.
The workflow spec creates the proper configuration out of it.
This should be cleaned up, the parsing inside the workflow should be removed, and we should use proper command line arguments. Anyhow, for now this is good enough and fully backward compatible.

Next I'll update this PR with:
- Rename the TPCCATracking just to CATracking, since it is merely an interface to the GPU tracking, which can also do TRD for example.
- Autodetect OpenGL during the build, and enable the event display by default if available @ktf 
- Add config settings to run the GPU tracking only partially, I.e. only do TPC tracking, start from TPC tracks and produce TRD tracks, etc.

shahor02 commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -74,7 +81,94 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
+      bool dump = false;
+      char gpuType[1024] = "CUDA";
+      GPUDisplayBackend* display = nullptr;
+      
```
@davidrohr @matthiasrichter: Could you add at once the GRP parameters parsing here?
````
#include "DataFormatsParameters/GRPObject.h"

const auto grp = o2::parameters::GRPObject::loadFrom("o2sim_grp.root");
if (grp) {
  LOG(INFO) << "Initializing run paramerers from GRP";
  solenoidBz *= grp->getL3Current()/30000.;
  continuous = grp->isDetContinuousReadOut(o2::detectors::DetID::TPC);
}
else {
  LOG(ERROR) << "Failed to initialize run parameters from GRP";
  // should we call fatal here?
}
````
I would keep at the moment hardcoded (default) grp file name (or add it as an option of the workflow, like in e.g. ``ITS/workflow/src/TrackerFlow.cxx``), eventually, this will be centrally substituted by the CCDB query or reading from the filename provided by the configurable names database.

davidrohr replied:
@shahor02 : Sure, will add it. Just one question: Where is the continuous mode set? During o2sim? I mean, this needs to be synchronized with the setting of "--TPCtriggered" in the digitizer-workflow.

shahor02 replied:
@davidrohr ``o2-sim`` just creates GRP with participating detectors and field data. Then ``o2-sim-digitizer-workflow`` updates the GRP according to readout options requested for each detector, i.e. whether the  ``grp->isDetContinuousReadOut(o2::detectors::DetID::TPC);`` returns ``false`` (``true``) depends exactly on the presence (absence) of ``--TPCtriggered`` in the digitization.

davidrohr commented:
@dberzano : I added some addional detection for OpenGL and GLEW here: https://github.com/AliceO2Group/AliceO2/pull/2038/commits/c40ae7743371e33ee608b749bc51f87281228778#diff-6962d261f583223f49f8b58a680bbb01

OpenGL is basically already detected by GLFW, but I need separate detection for the support library GLU, which is done by find_package(OpenGL).

shahor02 commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -74,7 +82,104 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
+
+      // Default Settings
+      float solenoidBz = -5.00668; // B-field
+      float refX = 1000.;          // transport tracks to this x after tracking, >500 for disabling
```
@davidrohr: just noticed: is there a reason not use by default refX=83, expected in the matching?

davidrohr replied:
no particular reason, i can change that.

davidrohr replied:
done

davidrohr commented:
Status update:
- auto detection of OpenGL and compilation of the standalone event display is in and working in the C.I. (still disabled for MacOS) @ktf 
- Renaming and configuration changes are done
- What is now missing is an interface to set input and output pointers, since right now there is only one function with ptr to TPC clusters / TPC tracks, need to add compressed clusters, and possibility for TRD tracklets / tracks in the future.

davidrohr commented:
Shouldn't be merged before #2037, since it now contains the other PR as I couldn't rebase it properly without the other one in.
Configuration for inputs / output is no more or less finalized. The TPC workflow is adapted such that it behaves as before, but now it can be extended, first to store compressed clusters, and then also more generally everything is ready for TRD tracking in a workflow.

davidrohr commented:
@MichaelLettrich : Ping: this PR will contain creation of entropy-compressed clusters, basically what will be the input for your entropy compression.
The output is currently the ROOT-serializable container `CompressedClusters`.
It is a Struct of Arrays, with several C arrays, and some counters for the size of the arrays. You can learn from the comments for the ROOT serialization which array is how large.
You can basically compress all these arrays individually with different ANS tables, with some exceptions:
- rowDiffA and sliceLegDiffA should be encoded together.
- sigmaPadA and sigmaTimeA should be encoded together.
- sigmaPadU and sigmaTimeU should be encoded together.
- qTotA and qMaxA should be encoded together.
- qTotU and qMaxU should be encoded together.
With together, I mean that e.g. out of two 8-bit fields, you should create a 16 bit field, and encode that, e.g. qCombinedA[i] = (qTotA[i] << n) | qMaxA[i].

matthiasrichter commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -73,9 +86,132 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
+      if (opt && *opt) {
+        printf("Received options %s\n", opt);
+        const char* optPtr = opt;
+        while (optPtr && *optPtr) {
```
An example of such a parser can be found in the FairMQ channel options parsing
https://github.com/FairRootGroup/FairMQ/blob/3582091b1c76cbe689b8e66813ea71dd5ff55bc5/fairmq/options/FairMQSuboptParser.h#L46
https://github.com/FairRootGroup/FairMQ/blob/ee24144d6158c00a7bc2293e28ca8e4a0c36e9f6/fairmq/options/FairMQSuboptParser.cxx#L32

davidrohr replied:
again agreed, just moved it here as first step and in order to be compatible.

shahor02 replied:
> key[=value] list, e.g. `refX=83,cont`. That would have the advantage to directly use system tool 

The refX is the reconstruction option (its counterpart in the matching code is at the moment hardcoded), should be eventually extracted from Sandro's ConfigurableParam-derived class (stored in ccdb and can be overridden either from the ini file or from command line, e.g.  ``--configKeyValues MatchingParams.mRefX=...``.
Also, since the field and cont/triggered RO mode must correspond to simulation and digitization (hence will be always extracted from the GRP), there is no much sense in looking for them in the options after having parsed from the GRP.

davidrohr commented on Detectors/TPC/reconstruction/src/GPUCATracking.cxx:
```diff
@@ -0,0 +1,253 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
```
quite sure I did, perhaps after the rebasing it does now longer recognize this?

davidrohr commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -279,6 +419,14 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
       }
 
+      // TODO - Process Compressed Clusters Output
+      const o2::tpc::CompressedClusters* compressedClusters = ptrs.compressedClusters; // This is a ROOT-serializable container with compressed TPC clusters
```
yes, I mean sending

davidrohr commented on GPU/GPUTracking/Standalone/utils/bitfield.h:
```diff
@@ -50,9 +50,18 @@ class bitfield
     return *this;
   }
   void set(S v) { bits = v; }
+  void set(T v) { bits = (S)v; }
```
I can change the casts, and check the size.
I want T to be part of the class, as it kind of defines what values are allowed.

matthiasrichter commented:
@davidrohr I read trough your initial comment, so ok, we can take this as an intermediate version, but probably one can spent the effort to create a separate helper class for the generation of the configuration, and allow also for the comma-separated list. Then we can deprecate the blank as separation and drop it later.

davidrohr commented on GPU/GPUTracking/Standalone/utils/bitfield.h:
```diff
@@ -50,9 +50,18 @@ class bitfield
+  void set(T v, Args... args)
+  {
+    this->set(args...);
+    *this |= v;
```
it's all inline, so it'll go away, parameters are basically integers anyway. Anyhow, I can make the paramters const ref, then it might be cleaner.

davidrohr commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -73,9 +86,132 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
+            printf("Dumping of input events enabled\n");
+          }
+#ifdef BUILD_EVENT_DISPLAY
+          else if (strncmp(optPtr, "display", optLen) == 0) {
```
I think the string parsing should be moved out of the spec anyway, just moved it here for now to stay compatible. I can for sure move the `#ifdef` inside the `else if` and throw an error otherwise.

davidrohr commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -73,9 +86,132 @@ DataProcessorSpec getCATrackerSpec(bool processMC, std::vector<int> const& input
+      }
+
+      // Parse the config string
+      const char* opt = options.c_str();
```
Agreed, but as I said, this was just a temporary solution to clean up the interface. I'd do away with this string parsing here completely, and use proper command line options of the workflow. How this is organized, I don't really care, but I wouldn't necessarily put it inside `GPUO2InterfaceConfiguration`. But I think this can be cleaned up later. Meaning of this PR was to define and clean up the interface between the spec and GPUCAReconstruction.

davidrohr commented:
@matthiasrichter : So I suggest I clean up the things you mentioned, and I also have one fix by myself.
Also, I am adding one more patch that exposes more of the compression options in the spec.
Then, I'd merge this (for the time being, it's fully backward compatible).

Next, you can go on to clean up and improve the workflow itself? Also, we can then discuss tomorrow how we enabled the split processing with two instances of GPUCATracking running consecutive steps.

davidrohr commented on Detectors/TPC/reconstruction/include/TPCReconstruction/DigitalCurrentClusterIntegrator.h:
```diff
@@ -8,7 +8,7 @@
 // or submit itself to any jurisdiction.
 
-/// \file TPCCATracking.h
+/// \file GPUCATracking.h
```
ah, ok, this was the wrong name before, and my sed touched it... :)

davidrohr commented:
@matthiasrichter : Comments should be addressed, I fixed also my other problem, and the dependent PR is merged. I have added some more relevant config options to the workflow spec. Guess I am done now. Removing the [WIP]. Talk to you tomorrow. This can be merged.
I'll leave it up to you to actually merge it, in case you are OK with the new version and CI passes now.
(I assume the MacOS C.I. will fail again due to the unrelated problem with the ccdb test case.)

sawenzel commented:
@ktf, @dberzano : This PR is in pending O2/linux build status for the last 5hours. I guess something is wrong.

ktf commented:
The builder was swamped by the zombies... I removed the machine from the cluster for now. @costing any change we can reboot pcalienstorage?

ktf commented:
For the record, the issue with zombies is understood and fixed.

More details at:

https://alice.its.cern.ch/jira/browse/O2-720
